tests/api_fastapi/execution_api/routes/test_task_instance.py

```diff
@@ -0,0 +1,192 @@
+        marked as running.
+        """
+        test_client = client(apps="execution")
+        ti = create_task_instance(task_id="test_ti_update_state_to_running", state=State.QUEUED)
+
+        session.commit()
+
+        response = test_client.patch(
+            f"/execution/task_instance/{ti.id}/state",
+            json={
+                "state": "running",
+                "hostname": "random-hostname",
+                "unixname": "random-unixname",
+                "pid": 100,
+                "start_date": "2024-10-31T12:00:00Z",
+            },
+        )
+
+        assert response.status_code == 204
+        assert response.text == ""
+
+        assert session.scalar(select(TaskInstance.state).where(TaskInstance.id == ti.id)) == State.RUNNING
+
+    def test_ti_update_state_conflict_if_not_queued(self, client, session, create_task_instance):
+        """
+        Test that a 409 error is returned when the Task Instance is not in a state where it can be marked as
+        running. In this case, the Task Instance is first in NONE state so it cannot be marked as running.
+        """
+        ti = create_task_instance(
+            task_id="test_ti_update_state_conflict_if_not_queued",
+            state=State.NONE,
+        )
+        session.commit()
+        test_client = client(apps="execution")
+
+        response = test_client.patch(
+            f"/execution/task_instance/{ti.id}/state",
+            json={
+                "state": "running",
+                "hostname": "random-hostname",
+                "unixname": "random-unixname",
+                "pid": 100,
+                "start_date": "2024-10-31T12:00:00Z",
+            },
+        )
+
+        assert response.status_code == 409
+        assert response.json() == {
+            "detail": {
+                "message": "TI was not in a state where it could be marked as running",
+                "previous_state": State.NONE,
+                "reason": "invalid_state",
+            }
+        }
+
+        assert session.scalar(select(TaskInstance.state).where(TaskInstance.id == ti.id)) == State.NONE
+
+    @pytest.mark.parametrize(
+        "state, end_date, expected_state",
+        [
+            (State.SUCCESS, "2024-10-31T12:00:00Z", State.SUCCESS),
+            (State.FAILED, "2024-10-31T12:05:00Z", State.FAILED),
+            (State.SKIPPED, "2024-10-31T12:05:00Z", State.SKIPPED),
+            (State.SKIPPED, None, State.SKIPPED),
+            (State.UPSTREAM_FAILED, None, State.UPSTREAM_FAILED),
+            (State.REMOVED, None, State.REMOVED),
+        ],
+    )
+    def test_ti_update_state_to_terminal(
+        self, client, session, create_task_instance, state, end_date, expected_state
+    ):
+        ti = create_task_instance(
+            task_id="test_ti_update_state_to_terminal",
+            state=State.RUNNING,
+        )
+        session.commit()
```

Comment:
Do we need the commit here? Should we be passing the session in to `create_task_instance` too?

Reply:
Updated